### PR TITLE
KNOX-2555 - Add a Token Generation Page for Acquiring JWT Tokens

### PR DESCRIPTION
**Pull request by lmccay**
Change-Id: I888cbfc43df1f7c6d706e3391ea7225acd70846f

## What changes were proposed in this pull request?

This patch provides a new Token Generation application that can be hosted in any topology just as any Knox application. It will be added to the homepage.xml topology initially.

In addition, this application requires the KnoxToken service to be added to the same topology for acquiring tokens.

It also allows the user to select a lifespan for the JWT token in days.
This requires a change to the KnoxToken service itself to accept a lifespan query param.
The requested lifespan must be less than or equal to the configured TTL for the KnoxService.

## How was this patch tested?

Manually tested from homepage->general proxy information->Token Generation

Resulting JWT token used as an http password using enhancement from KNOX-2539.
Used it as an authorization bearer token.

Ran existing unit tests.

### CODE REVIEW

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java [hunk 1]
```diff
@@ -91,6 +92,7 @@
   static final String RENEW_PATH = "/renew";
   static final String REVOKE_PATH = "/revoke";
   private static final String TARGET_ENDPOINT_PULIC_CERT_PEM = "knox.token.target.endpoint.cert.pem";
+  private static final long MILLIS_IN_DAY = 86400000L;
```
Comment (smolnar82) [inline gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java:95]: nit: TimeUnit.DAYS.toMillis()

File: gateway-release/home/conf/topologies/homepage.xml [hunk 2]
```diff
@@ -70,6 +75,20 @@
...
+   <service>
+      <role>KNOXTOKEN</role>
+      <param>
+         <name>knox.token.ttl</name>
+         <value>604800000</value>
+      </param>
+      <param>
+         <name>knox.token.target.url</name>
+         <value>cdp-proxy-token/</value>
+      </param>
```
Comment (smolnar82) [inline gateway-release/home/conf/topologies/homepage.xml:87]: nit: The target URL should have been removed

**Review by smolnar82 — APPROVED** [review]
LGTM +1

As discussed this is going to be re-written as an Angular2 application using typescript to be consistent with the rest of our UI apps.

**Comment by smolnar82** [comment]
Cloned your repo, built and tested. Works like a charm...ship it :)

Screen Shot 2021-03-19 at 10 11 38 AM

File: gateway-release/home/conf/topologies/homepage.xml (hunk 2 above)
Comment (lmccay) [reply gateway-release/home/conf/topologies/homepage.xml]: I've removed the 'cdp-' to make more generic. Thanks for the catch.

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java (hunk 1 above)
Comment (lmccay) [reply gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java:95]: Nice - can follow up in another commit. I have other changes in mind as well anyway.

**Pull request merged**